src/pure_mvreg.erl

```diff
@@ -0,0 +1,176 @@
+
+-export_type([pure_mvreg/0, pure_mvreg_op/0]).
+
+-opaque pure_mvreg() :: {?TYPE, payload()}.
+-type payload() :: {pure_type:polog(), []}.
+-type pure_mvreg_op() :: {set, string()}.
+
+%% @doc Create a new, empty `pure_mvreg()'
+-spec new() -> pure_mvreg().
+new() ->
+    {?TYPE, {orddict:new(), []}}.
+
+%% @doc Create a new, empty `pure_mvreg()'
+-spec new([term()]) -> pure_mvreg().
+new([]) ->
+    new().
+
+%% @doc Check redundancy `pure_mvreg()'
+%% Called in remove_redundant().
+-spec redundant({pure_type:id(), pure_mvreg_op()}, {pure_type:id(), pure_mvreg_op()}) ->
+    integer().
+redundant({VV1, {_Op1, _Str1}}, {VV2, {_Op2, _Str2}}) ->
+    case pure_trcb:happened_before(VV1, VV2) of
+        true ->
+            1;
+        false ->
+            0
+    end.
+
+%% @doc Removes redundant operations from POLog of `pure_mvreg()'
+-spec remove_redundant_polog({pure_type:id(), pure_mvreg_op()}, pure_mvreg()) -> {boolean(), pure_mvreg()}.
+remove_redundant_polog({VV1, Op}, {?TYPE, {POLog0, MVReg}}) ->
+    case orddict:is_empty(POLog0) of
+        true ->
+            {true, {?TYPE, {POLog0, MVReg}}};
+        false ->
+            {POLog1, Add1} = orddict:fold(
+                fun(Key, Value, {Acc, Add}) ->
+                    case redundant({Key, {set, Value}}, {VV1, Op}) of
+                        0 ->
+                            {orddict:store(Key, Value, Acc), Add};
+                        1 ->
+                            {Acc, Add}
+                    end
+                end,
+                {orddict:new(), true},
+                POLog0
+            ),
+            {Add1, {?TYPE, {POLog1, MVReg}}}
+    end.
+
+%% @doc Removes redundant operations from crystal of `pure_mvreg()'
+%% No crystalisation for this one.
+-spec remove_redundant_crystal({pure_type:id(), pure_mvreg_op()}, pure_mvreg()) -> {boolean(), pure_mvreg()}.
+remove_redundant_crystal({_VV1, _X}, {?TYPE, {POLog0, MVReg}}) ->
+    {true, {?TYPE, {POLog0, MVReg}}}.
+
+%% @doc Checks stable operations and remove them from POLog of `pure_mvreg()'
+-spec check_stability(pure_type:id(), pure_mvreg()) -> pure_mvreg().
+check_stability(_StableVV, {?TYPE, {POLog0, MVReg0}}) ->
+    {?TYPE, {POLog0, MVReg0}}.
+
+%% @doc Update a `pure_mvreg()'.
+-spec mutate(pure_mvreg_op(), pure_type:id(), pure_mvreg()) ->
+    {ok, pure_mvreg()}.
+mutate({Op, Str}, VV, {?TYPE, {POLog, PureMVReg}}) ->
+    {_Add, {?TYPE, {POLog0, PureMVReg0}}} = pure_polog:remove_redundant({VV, {Op, Str}}, {?TYPE, {POLog, PureMVReg}}),
+    {ok, {?TYPE, {orddict:store(VV, Str, POLog0), PureMVReg0}}}.
+
+%% @doc Returns the value of the `pure_mvreg()'.
+%%      This value is a a boolean value in the `pure_mvreg()'.
+-spec query(pure_mvreg()) -> [string()].
+query({?TYPE, {POLog0, _PureMVReg0}}) ->
+    case orddict:is_empty(POLog0) of
+        true ->
+            [];
+        false ->
+            [Str || {_Key, Str} <- orddict:to_list(POLog0)]
+    end.
+
```

Comment:
I think that in case the same value is written concurrently it will be reported many times in the MVReg. Should we be careful in the query so as to not report multiple identical elements. (I think I discussed with Georges that the identical elements probably need to all be stored, since they can be independently overwritten.)